### ISSUE TITLE
Ragel 7.0.0.9 can't process smtp_received_parser.rl

### ISSUE THREAD

**Issue by hexum** [vstakhov/rspamd#1396]
### Classification (Please choose *one* option):

* [ ] Crash/Hang/Data loss
* [ ] WebUI/Usability
* [ ] Serious bug
* [x] Other bug
* [ ] Feature
* [ ] Enhancement

### Reproducibility (Please choose *one* option):

* [X] Always
* [ ] Sometimes
* [ ] Rarely
* [ ] Unable
* [ ] I didn’t try
* [ ] Not applicable

### Rspamd version:

GIT NEW branch -->
   repository:               https://github.com/vstakhov/rspamd.git
   at the commit:            6c8b19b9422f6fb4b776a59617750d53a48da5f1

### Operation system, CPU, memory and environment:

Gentoo linux
=dev-util/ragel-7.0.0.9

### Description (Please provide a descriptive summary of the issue):

Ragel failed on ragel/smtp_received_parser.rl processing

### Compile errors (if any):

```
[ 44%] [RAGEL][ragel_smtp_received] Compiling state machine with Ragel 7.0.0.9
cd /var/tmp/portage/mail-filter/rspamd-9999/work/rspamd-9999/src && /usr/bin/ragel -T1 -o/var/tmp/portage/mail-filter/rspamd-9999/work/rspamd-9999_build/src/smtp_received_parser.rl.c /var/tmp/portage/mail-filter/rspamd-9999/work/rspamd-9999/src/ragel/smtp_received_parser.rl
<internal>:4:3: fsm "WSP" previously defined
<internal>:5:3: fsm "CRLF" previously defined
<internal>:6:3: fsm "DQUOTE" previously defined
<internal>:9:3: fsm "atext" previously defined
<internal>:13:3: fsm "dtext" previously defined
<internal>:15:3: fsm "ctext" previously defined
<internal>:17:3: fsm "dcontent" previously defined
<internal>:18:3: fsm "Let_dig" previously defined
<internal>:19:3: fsm "Ldh_str" previously defined
<internal>:21:3: fsm "quoted_pairSMTP" previously defined
<internal>:22:3: fsm "qtextSMTP" previously defined
<internal>:23:3: fsm "Atom" previously defined
<internal>:24:3: fsm "Dot_string" previously defined
<internal>:25:3: fsm "dot_atom_text" previously defined
<internal>:27:3: fsm "FWS" previously defined
<internal>:7:3: fsm "Snum" previously defined
<internal>:8:3: fsm "IPv4_address_literal" previously defined
<internal>:9:3: fsm "IPv6_hex" previously defined
<internal>:10:3: fsm "IPv6_full" previously defined
<internal>:11:3: fsm "IPv6_comp" previously defined
<internal>:13:3: fsm "IPv6v4_full" previously defined
<internal>:14:3: fsm "IPv6v4_comp" previously defined
<internal>:17:3: fsm "IPv6_addr" previously defined
<internal>:18:3: fsm "IPv6_address_literal" previously defined
<internal>:4:3: fsm "WSP" previously defined
<internal>:5:3: fsm "CRLF" previously defined
<internal>:6:3: fsm "DQUOTE" previously defined
<internal>:9:3: fsm "atext" previously defined
<internal>:13:3: fsm "dtext" previously defined
<internal>:15:3: fsm "ctext" previously defined
<internal>:17:3: fsm "dcontent" previously defined
<internal>:18:3: fsm "Let_dig" previously defined
<internal>:19:3: fsm "Ldh_str" previously defined
<internal>:21:3: fsm "quoted_pairSMTP" previously defined
<internal>:22:3: fsm "qtextSMTP" previously defined
<internal>:23:3: fsm "Atom" previously defined
<internal>:24:3: fsm "Dot_string" previously defined
<internal>:25:3: fsm "dot_atom_text" previously defined
<internal>:27:3: fsm "FWS" previously defined
-:1:1: <input>:1:1: parse error^@%s:%ld:
```

### Steps to reproduce:

Compile master.
```
cmake --no-warn-unused-cli -C /var/tmp/portage/mail-filter/rspamd-9999/work/rspamd-9999_build/gentoo_common_config.cmake -G Unix Makefiles -DCMAKE_INSTALL_PREFIX=/usr -DCONFDIR=/etc/rspamd -DRUNDIR=/var/run/rspamd -DDBDIR=/var/lib/rspamd
 -DLOGDIR=/var/log/rspamd -DENABLE_LUAJIT=ON -DENABLE_FANN=OFF -DENABLE_PCRE2=ON -DENABLE_JEMALLOC=ON -DENABLE_GD=ON -DCMAKE_BUILD_TYPE=Gentoo -DCMAKE_INSTALL_DO_STRIP=OFF -DCMAKE_USER_MAKE_RULES_OVERRIDE=/var/tmp/portage/mail-filter/rsp
amd-9999/work/rspamd-9999_build/gentoo_rules.cmake -DCMAKE_TOOLCHAIN_FILE=/var/tmp/portage/mail-filter/rspamd-9999/work/rspamd-9999_build/gentoo_toolchain.cmake  /var/tmp/portage/mail-filter/rspamd-9999/work/rspamd-9999
```

### Expected results:

Compilation done.

### Actual results:

Compilation failed.

### Debugging information (see details [here] (https://rspamd.com/doc/faq.html#how-to-figure-out-why-rspamd-process-crashed)):

Build time error.

### Configuration:

`[ebuild  N    *] mail-filter/rspamd-9999  USE="gd jemalloc jit libressl pcre2 -fann"`

### Additional information:

Full compilation log attached.
[build.log.zip](https://github.com/vstakhov/rspamd/files/752585/build.log.zip)

**Comment by vstakhov** [comment]
This is not Rspamd bug as far as I see. You should contact ragel developers about it. ⁣

**Comment by hexum** [comment]
Master was built
OpenSC-0.16.0-170-g3635dbe7
Origainal bug with key upload is still reproducible.
Commands rewritten to automate testing by copy-paste.

**Comment by rtakamura** [comment]
I get the same error even in my environment.
CentOS 7.3 64 bit, Ragel 7.0.0.9

make log
- -
[44%] [RAGEL] [ragel_smtp_received] Compiling state machine with Ragel 7.0.0.9
<Internal>: 4: 3: fsm "WSP" previously defined
<Internal>: 5: 3: fsm "CRLF" previously defined
.......
-: 1: 1: <input>: 1: 1: parse error% s:% ld:
Make [2]: *** [src / smtp_received_parser.rl.c] Error 1
Make [1]: *** [src / CMakeFiles / rspamd-server.dir / all] Error 2
Make: *** [all] Error 2
- -

Ragel version:
Ragel-7.0.0.9-2.el7.x86_64 by epel

**Comment by vstakhov** [comment]
Ragel 7.0 is not the stable version. It is obviously broken here. No sane 
distros use this version. Hence, you should open bug report to the ragel 
project itself (or centos msintainer). There is clearly nothing to do on 
Rspamd side. And there is a workaround posted here. Or you can use packages 
provided.

**Comment by FelixSchwarz** [comment]
Adrian Thurston (ragel's main author) just added fixes for this (duplicate includes, relative include paths). I did not have time to verify this but the very latest ragel git is supposed to build rspamd again.

**Comment by LorbusChris** [comment]
To test this, I just tried to build a canary version of ragel from git master, but the build is failing:
```
aapl/Makefile.am:1: error: 'pkgpkginclude_HEADERS' is used but 'pkgpkgincludedir' is undefined
src/Makefile.am: installing './depcomp'
src/Makefile.am:25: error: 'nodist_pkgpkginclude_HEADERS' is used but 'pkgpkgincludedir' is undefined
src/Makefile.am:21: error: 'pkgpkginclude_HEADERS' is used but 'pkgpkgincludedir' is undefined
autoreconf: automake failed with exit status: 1
```

**Comment by vstakhov** [comment]
The question is why this unstable version is used in Centos/Fedora?

**Comment by FelixSchwarz** [comment]
Yeah, I think that was an unfortunate decision (my PoV). I guess the Fedora maintainer tried to get the (ragel) build fixed on some archs with the new version and maybe he was not aware of the unusual versioning scheme. Still that boat has sailed now but I found ragel's author pretty responsive so I recommend contacting him on the ragel dev list.

Once it is fixed to build rspamd again, you should ask for a new ragel dev release. That can be integrated in Fedora EPEL pretty easily.

**Comment by LorbusChris** [comment]
@FelixSchwarz @vstakhov 
See here for a ragel-compat spec:
https://src.fedoraproject.org/rpms/ragel/pull-request/2

**Comment by FelixSchwarz** [comment]
@LorbusChris I just checked and I was able to build colm+ragel+rspamd from git just fine on CentOS 7. I'll do some more checking but I think ragel git should work fine.

Just for reference I used these versions:

colm:

     commit 83e3cbcb25199888787e6d87c3bfa00cb0947462
     Author: Adrian Thurston <thurston@colm.net>
     Date:   Sun Jul 30 11:52:03 2017 -0400
     
         organization in reduce code generation

ragel:

     commit 8c6c43635c21f78a89ef115581bc9ffb2c22d300
     Author: Adrian Thurston <thurston@colm.net>
     Date:   Sun Sep 24 16:24:27 2017 -0400
     
         cntrl inclues 127, updated docs
         
          Caught by Rinat Ibragimov.


rspamd (git d0106139487f80bc4fcc926d7329f3e0aada234d):

    commit d0106139487f80bc4fcc926d7329f3e0aada234d
    Author: Vsevolod Stakhov <vsevolod@highsecure.ru>
    Date:   Wed Sep 27 20:44:56 2017 +0100
    
        [Fix] Check for magic when checking for an archive
        
        Issue: #1848
        Closes: #1848
        MFH: 1.6

**Comment by FelixSchwarz** [comment]
Just a final hint: If you use ragel git you also need ragel colm.

@LorbusChris: Probably you tried to rebuild ragel with a spec from EPEL which contains a sed expression which generates "pkgpkginclude_HEADERS". That sed call should be removed.

Other than that rspamd with ragel/colm git works fine for me so I think this bug should be closed as this is not a rspamd problem.
